fs/unpacker.go

```diff
@@ -49,26 +50,63 @@ func NewLayerArchive() Archive {
 	return &layerArchive{}
 }
 
-func (la *layerArchive) Apply(ctx context.Context, root string, r io.Reader, opts ...archive.ApplyOpt) (int64, error) {
+// wrapReader uses a TeeReader to feed content into verifier (compressed or uncompressed).
+// If verification is disabled, this is a no-op.
+func wrapReader(r io.Reader, verifier *layerVerifier, compressed bool) io.Reader {
+	if verifier != nil {
+		if compressed {
+			r = io.TeeReader(r, verifier.compressed)
+		} else {
+			r = io.TeeReader(r, verifier.uncompressed)
+		}
+	}
+	return r
+}
+
+func (la *layerArchive) Apply(ctx context.Context, root string, r io.Reader, verifier *layerVerifier, opts ...archive.ApplyOpt) (int64, error) {
 	// we use containerd implementation here
 	// decompress first and then apply
+	r = wrapReader(r, verifier, true)
 	decompressReader, err := compression.DecompressStream(r)
 	if err != nil {
 		return 0, fmt.Errorf("cannot decompress the stream: %w", err)
 	}
 	defer decompressReader.Close()
-	return archive.Apply(ctx, root, decompressReader, opts...)
+
+	r = wrapReader(decompressReader, verifier, false)
+	n, err := archive.Apply(ctx, root, r, opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	if verifier != nil {
```

Comment:
Instead of a bunch of nil checks around, can we add a concept of a no-op verifier that just always returns true? That way, we can guarantee it's never nil and reduce the number of special cases?